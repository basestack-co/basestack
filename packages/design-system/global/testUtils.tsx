import { type RenderOptions, render } from "@testing-library/react";
import type React from "react";
// Styles
import { ThemeProvider } from "styled-components";
import theme from "../theme/lightTheme";

interface Props {
  children: React.ReactNode;
}

const WithThemeProvider = ({ children }: Props) => (
  <ThemeProvider theme={theme}>{children}</ThemeProvider>
);

const renderWithTheme = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, "wrapper">,
) => render(ui, { wrapper: WithThemeProvider, ...options });

export { renderWithTheme };
