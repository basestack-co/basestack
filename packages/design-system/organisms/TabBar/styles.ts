import styled from "styled-components";
import { rem } from "polished";

export const Container = styled.nav`
  position: sticky;
  bottom: 0;
  z-index: ${({ theme }) => theme.zIndex.tabBar};
  display: flex;
  align-items: center;
  flex-shrink: 0;
  background-color: ${({ theme }) => theme.colors.white};
  box-shadow:
    rgb(0 0 0 / 10%) 0 0 3px 0,
    rgb(0 0 0 / 6%) 0 0 2px 0;
  height: ${rem("64px")};
  margin-top: auto;
`;

export const List = styled.ul`
  display: flex;
  align-items: center;
  margin: 0 auto;
`;

export const ListItem = styled.li`
  padding: 0 ${({ theme }) => theme.spacing.s2};

  @media screen and (min-width: 380px) {
    padding: 0 ${({ theme }) => theme.spacing.s3};
  }

  @media screen and ${({ theme }) => theme.device.min.sm} {
    padding: 0 ${({ theme }) => theme.spacing.s4};
  }
`;

export const StyledButton = styled.a<{ isActive: boolean }>`
  text-decoration: none;
  cursor: pointer;
  display: flex;
  align-items: center;
  justify-content: center;
  font-size: ${rem("14px")};
  font-weight: 500;
  height: ${rem("64px")};
  padding: 0 ${({ theme }) => theme.spacing.s3};
  color: ${({ theme, isActive }) =>
    isActive ? theme.colors.primary : theme.colors.black};
  border-top: ${({ theme, isActive }) =>
    `3px solid ${isActive ? theme.colors.primary : theme.colors.white}`};
`;
