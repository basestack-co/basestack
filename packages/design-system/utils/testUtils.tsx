import defaultTheme from "@basestack/design-system/theme/lightTheme";
import isPropValid from "@emotion/is-prop-valid";
import { RenderResult, render } from "@testing-library/react";
import React from "react";
import { StyleSheetManager, ThemeProvider } from "styled-components";

const Provider = ({
  children,
  theme = defaultTheme,
}: {
  children: React.ReactNode;
  theme: any;
}) => (
  <ThemeProvider theme={theme}>
    <StyleSheetManager shouldForwardProp={isPropValid}>
      {children}
    </StyleSheetManager>
  </ThemeProvider>
);

export const renderWithTheme = (
  ui: React.ReactElement,
  options?: any,
): RenderResult => render(ui, { wrapper: Provider, ...options });
