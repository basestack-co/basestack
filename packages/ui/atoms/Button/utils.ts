import { Size, Target, Type, Variant } from "@/components/atoms/Button/types";
import { css, DefaultTheme } from "styled-components";
import { darken, rem, transparentize } from "polished";

export const handleButtonSize = (size?: Size) => {
  switch (size) {
    case "large":
      return css`
        height: ${rem("60px")};
        font-size: ${rem("18px")};
        line-height: ${rem("25px")};
        padding: 0 ${rem("40px")};
      `;
    case "medium":
      return css`
        height: ${rem("54px")};
        font-size: ${rem("16px")};
        line-height: ${rem("22px")};
        padding: 0 ${rem("34px")};
      `;
    case "normal":
      return css`
        height: ${rem("48px")};
        font-size: ${rem("16px")};
        line-height: ${rem("22px")};
        padding: 0 ${rem("30px")};
      `;
    default:
    case "small":
      return css`
        height: ${rem("46px")};
        font-size: ${rem("14px")};
        line-height: ${rem("19px")};
        padding: 0 ${rem("24px")};
      `;
    case "xSmall":
      return css`
        height: ${rem("42px")};
        font-size: ${rem("14px")};
        line-height: ${rem("19px")};
        padding: 0 ${rem("20px")};
      `;
  }
};

export const handleButtonsColors = (theme: DefaultTheme) => ({
  primary: {
    outline: {
      color: theme.colors.primary,
      borderColor: theme.colors.primary,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.white,
      borderColor: "transparent",
      backgroundColor: theme.colors.primary,
    },
  },
  secondary: {
    outline: {
      color: theme.colors.secondary,
      borderColor: theme.colors.secondary,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.white,
      borderColor: "transparent",
      backgroundColor: theme.colors.secondary,
    },
  },
  tertiary: {
    outline: {
      color: theme.colors.gray3,
      borderColor: theme.colors.gray3,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.primary,
      borderColor: "transparent",
      backgroundColor: theme.colors.gray6,
    },
  },
  neutral: {
    outline: {
      color: theme.colors.gray2,
      borderColor: theme.colors.gray2,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.primary,
      borderColor: "transparent",
      backgroundColor: "transparent",
    },
  },
  success: {
    outline: {
      color: theme.colors.success,
      borderColor: theme.colors.success,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.white,
      borderColor: "transparent",
      backgroundColor: theme.colors.success,
    },
  },
  warning: {
    outline: {
      color: theme.colors.warning,
      borderColor: theme.colors.warning,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.white,
      borderColor: "transparent",
      backgroundColor: theme.colors.warning,
    },
  },
  danger: {
    outline: {
      color: theme.colors.error,
      borderColor: theme.colors.error,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.white,
      borderColor: "transparent",
      backgroundColor: theme.colors.error,
    },
  },
});

export const darkenColors = (
  theme: DefaultTheme,
  amount: number,
  type: Type,
  variant: Variant,
  target: Target
) => {
  return darken(
    amount || 0.1,
    handleButtonsColors(theme)[type || "primary"][variant || "filled"][target]
  );
};

export const lightenColors = (
  theme: DefaultTheme,
  amount: number,
  type: Type,
  variant: Variant,
  target: Target
) => {
  return transparentize(
    amount || 0.6,
    handleButtonsColors(theme)[type || "primary"][variant || "filled"][target]
  );
};
