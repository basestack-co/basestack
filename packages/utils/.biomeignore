# Node dependencies
node_modules

# Next.js build output
.next
out

# Environment files
.env
.env.*
!.env.example

# Logs
*.log
npm-debug.log*
yarn-debug.log*
yarn-error.log*
pnpm-debug.log*

# OS / editor files
.DS_Store
Thumbs.db

# Package manager cache
.pnp
.pnp.js
.pnp.cjs

# Test coverage
coverage

# Build artifacts
dist
build

# Temporary files
.tmp
temp