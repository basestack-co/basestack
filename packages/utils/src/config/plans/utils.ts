export const getAppMode = (mode: string) => {
  return (
    ["development", "production", "local"].includes(mode) ? mode : "production"
  ) as "local" | "development" | "production";
};
