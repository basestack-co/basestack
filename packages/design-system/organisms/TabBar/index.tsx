import { memo } from "react";
import { useTheme } from "styled-components";
import { IconButton, Icon } from "../../atoms";
import { Container, List, ListItem, StyledButton } from "./styles";
import { useMediaQuery } from "@basestack/hooks";

export interface ButtonProps {
  icon: string;
  text: string;
  isActive: boolean;
  href: string;
}

export interface TabBarProps {
  /**
   * create flag callback
   */
  onCreateFlag: () => void;
  pathname: string;
}

const ButtonLink = ({ icon, text, isActive, href }: ButtonProps) => {
  const theme = useTheme();
  const isMediumDevice = useMediaQuery(theme.device.min.md);

  return (
    <StyledButton href={href} isActive={isActive}>
      <Icon
        icon={icon}
        size="medium"
        mr={isMediumDevice ? theme.spacing.s2 : 0}
        color={isActive ? theme.colors.primary : theme.colors.black}
      />
      {isMediumDevice && <span>{text}</span>}
    </StyledButton>
  );
};

const TabBar = ({ onCreateFlag, pathname }: TabBarProps) => {
  const theme = useTheme();

  return (
    <Container data-testid="tab-bar">
      <List>
        <ListItem>
          <ButtonLink
            href="/flags"
            isActive={pathname === "/flags"}
            icon="flag"
            text="Features"
          />
        </ListItem>
        <ListItem>
          <ButtonLink
            href="/settings/general"
            isActive={pathname ? pathname.includes("settings") : false}
            icon="settings"
            text="Settings"
          />
        </ListItem>
        <ListItem>
          <IconButton
            mx={theme.spacing.s3}
            variant="primary"
            size="large"
            icon="add"
            onClick={onCreateFlag}
          />
        </ListItem>
        <ListItem>
          <ButtonLink href="/" isActive={false} icon="flag" text="Features" />
        </ListItem>
        <ListItem>
          <ButtonLink
            href="/"
            isActive={false}
            icon="settings"
            text="Settings"
          />
        </ListItem>
      </List>
    </Container>
  );
};

export default memo(TabBar);
