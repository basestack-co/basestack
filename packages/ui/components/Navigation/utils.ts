// Types
import { Translate } from "next-translate";
import { NextRouter } from "next/router";
// Utils
import { config } from "@basestack/utils";
import { signOut } from "next-auth/react";

export const getInternalLinks = (t: Translate, formId: string) => {
  return [
    {
      icon: "flag",
      activeText: "Submissions",
      to: `/form/${formId}/submissions`,
      text: t("internal.submissions"),
    },
    {
      icon: "flag",
      activeText: "Setup",
      to: `/form/${formId}/setup`,
      text: t("internal.setup"),
    },
    {
      icon: "settings",
      activeText: "Settings",
      to: `/form/${formId}/settings/general`,
      text: t("internal.settings"),
    },
  ];
};

export const getExternalLinks = (t: Translate) => {
  return [
    {
      icon: "description",
      activeText: "Documentation",
      to: config.urls.docs.flags.base,
      isExternal: true,
      text: t("external.docs"),
    },
    {
      icon: "link",
      activeText: "Github",
      to: config.urls.repo,
      isExternal: true,
      text: t("external.github"),
    },
  ];
};

export const getAvatarDropdownList = (
  t: Translate,
  router: NextRouter,
  onCreateForm: () => void,
) => {
  return [
    {
      icon: "add_circle",
      text: t("create.form"),
      onClick: onCreateForm,
      separator: true,
    },
    {
      icon: "settings",
      text: t("dropdown.settings"),
      onClick: () =>
        router.push({
          pathname: "/user/profile/settings",
        }),
    },
    {
      icon: "logout",
      text: t("dropdown.logout"),
      onClick: signOut,
    },
  ];
};
