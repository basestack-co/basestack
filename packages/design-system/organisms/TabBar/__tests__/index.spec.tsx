import { cleanup } from "@testing-library/react";
import { renderWithTheme } from "../../../utils/testUtils";
import TabBar from "..";

describe("TabBar Organism tests", () => {
  beforeAll(() => {
    Object.defineProperty(window, "matchMedia", {
      value: jest.fn(() => {
        return {
          matches: true,
          addListener: jest.fn(),
          removeListener: jest.fn(),
        };
      }),
    });
  });

  afterEach(cleanup);

  test("should render TabBar correctly", () => {
    const { asFragment } = renderWithTheme(
      <TabBar onCreateFlag={jest.fn} pathname="/settings" />,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
