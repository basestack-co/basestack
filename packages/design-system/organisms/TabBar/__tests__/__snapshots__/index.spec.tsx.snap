// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabBar Organism tests should render TabBar correctly 1`] = `
<DocumentFragment>
  .c4 {
  height: 1.5rem;
  width: 1.5rem;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  margin-right: 0.5rem;
}

.c9 {
  height: 2rem;
  width: 2rem;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.c5 {
  color: #000000;
  font-size: 1.5rem;
}

.c7 {
  color: #276EF1;
  font-size: 1.5rem;
}

.c10 {
  color: #FFFFFF;
  font-size: 2rem;
}

.c8 {
  margin-left: 0.75rem;
  margin-right: 0.75rem;
  border: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  cursor: pointer;
  border-radius: 50%;
  -webkit-transition: background-color 0.1s ease-in-out,color 0.1s ease-in-out, border-color 0.1s ease-in-out;
  transition: background-color 0.1s ease-in-out,color 0.1s ease-in-out, border-color 0.1s ease-in-out;
  background-color: #276EF1;
  color: #FFFFFF;
  height: 2.5rem;
  width: 2.5rem;
}

.c8:hover:not(:active) {
  background-color: #0e55d7;
}

.c0 {
  position: -webkit-sticky;
  position: sticky;
  bottom: 0;
  z-index: 99;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  background-color: #FFFFFF;
  box-shadow: rgb(0 0 0 / 10%) 0 0 3px 0,rgb(0 0 0 / 6%) 0 0 2px 0;
  height: 4rem;
  margin-top: auto;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin: 0 auto;
}

.c2 {
  padding: 0 0.5rem;
}

.c3 {
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  font-size: 0.875rem;
  font-weight: 500;
  height: 4rem;
  padding: 0 0.75rem;
  color: #000000;
  border-top: 3px solid #FFFFFF;
}

.c6 {
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  font-size: 0.875rem;
  font-weight: 500;
  height: 4rem;
  padding: 0 0.75rem;
  color: #276EF1;
  border-top: 3px solid #276EF1;
}

@media screen and (min-width:380px) {
  .c2 {
    padding: 0 0.75rem;
  }
}

@media screen and (min-width:576px) {
  .c2 {
    padding: 0 1rem;
  }
}

<nav
    class="c0"
    data-testid="tab-bar"
  >
    <ul
      class="c1"
    >
      <li
        class="c2"
      >
        <a
          class="c3"
          href="/flags"
        >
          <div
            class="c4"
            data-testid="icon-container"
            size="24"
          >
            <span
              class="c5 material-symbols-sharp"
              color="#000000"
              data-testid="icon"
              size="24"
            >
              flag
            </span>
          </div>
          <span>
            Features
          </span>
        </a>
      </li>
      <li
        class="c2"
      >
        <a
          class="c6"
          href="/settings/general"
        >
          <div
            class="c4"
            data-testid="icon-container"
            size="24"
          >
            <span
              class="c7 material-symbols-sharp"
              color="#276EF1"
              data-testid="icon"
              size="24"
            >
              settings
            </span>
          </div>
          <span>
            Settings
          </span>
        </a>
      </li>
      <li
        class="c2"
      >
        <button
          class="c8"
        >
          <div
            class="c9"
            data-testid="icon-container"
            size="32"
          >
            <span
              class="c10 material-symbols-sharp"
              color="#FFFFFF"
              data-testid="icon"
              size="32"
            >
              add
            </span>
          </div>
        </button>
      </li>
      <li
        class="c2"
      >
        <a
          class="c3"
          href="/"
        >
          <div
            class="c4"
            data-testid="icon-container"
            size="24"
          >
            <span
              class="c5 material-symbols-sharp"
              color="#000000"
              data-testid="icon"
              size="24"
            >
              flag
            </span>
          </div>
          <span>
            Features
          </span>
        </a>
      </li>
      <li
        class="c2"
      >
        <a
          class="c3"
          href="/"
        >
          <div
            class="c4"
            data-testid="icon-container"
            size="24"
          >
            <span
              class="c5 material-symbols-sharp"
              color="#000000"
              data-testid="icon"
              size="24"
            >
              settings
            </span>
          </div>
          <span>
            Settings
          </span>
        </a>
      </li>
    </ul>
  </nav>
</DocumentFragment>
`;
